modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
+                        continue;
+                    }
+                }
+
+                // this index has matching downsample rounds and is read-only, so let's wait for an in-progress downsampling operation to
+                // succeed or trigger the corresponding one
+                // TODO add support for cancelling a current in-progress operation if another, later, round matches
+                DataStreamLifecycle.Downsampling.Round lastRound = downsamplingRounds.get(downsamplingRounds.size() - 1);
+
+                for (DataStreamLifecycle.Downsampling.Round round : downsamplingRounds) {
+                    // the downsample index name for each round is deterministic
+                    String downsampleIndexName = DownsampleConfig.generateDownsampleIndexName(
+                        DOWNSAMPLED_INDEX_PREFIX,
+                        backingIndex,
+                        round.config().getFixedInterval()
+                    );
+                    IndexMetadata downsampleIndexMeta = metadata.index(downsampleIndexName);
+                    if (downsampleIndexMeta != null) {
+                        IndexMetadata.DownsampleTaskStatus downsampleStatus = INDEX_DOWNSAMPLE_STATUS.get(
+                            downsampleIndexMeta.getSettings()
+                        );
+                        if (downsampleStatus.equals(UNKNOWN) && round.equals(lastRound)) {
+                            // target downsampling index exists and is not a downsampling index (name clash?)
+                            // we fail now but perhaps we should just randomise the name?
+                            errorStore.recordError(
+                                backingIndex.getIndex().getName(),
+                                new ResourceAlreadyExistsException(downsampleIndexMeta.getIndex().getName())
+                            );
+                        } else if (downsampleStatus.equals(STARTED)) {
+                            // we'll wait for this round to complete
+                            // TODO when cancelling support is added it'll go here, if there's a further round that can be started for
+                            // this index
+                            affectedIndices.add(backingIndex.getIndex());
+                            break;
+                        } else if (downsampleStatus.equals(SUCCESS)) {
+                            // at this point the source index is part of the data stream and the downsample index is complete but not
+                            // part of the data stream. we need to replace the source index with the downsample index in the data stream
+                            // and delete the source index.
+                            if (dataStream.getIndices().contains(downsampleIndexMeta.getIndex()) == false) {
+                                affectedIndices.add(backingIndex.getIndex());
+                                clusterStateChangesDeduplicator.executeOnce(
```

Comment:
I think having some debug logging here about what DSL is doing might be useful

Reply:
We have a `trace` log already - so the pattern is that we log in the method that gets executed inside the deduplicator (or every time the request actually goes through)

In this instance the logging is available here https://github.com/elastic/elasticsearch/pull/98609/files#diff-157b7d593f2d23d829609512ec2a6afeeeacda8a4ae31408e83861b747ef31bfR760